--- a/boost/thread/pthread/thread_data.hpp
+++ b/boost/thread/pthread/thread_data.hpp
@@ -57,8 +57,12 @@ namespace boost
 #else
           std::size_t page_size = ::sysconf( _SC_PAGESIZE);
 #endif
-#if PTHREAD_STACK_MIN > 0
-          if (size<PTHREAD_STACK_MIN) size=PTHREAD_STACK_MIN;
+#ifdef PTHREAD_STACK_MIN
+#if defined(__GLIBC__) && defined(_GNU_SOURCE) && PTHREAD_STACK_MIN == 0
+          std::size_t min_stack = sysconf(_SC_THREAD_STACK_MIN);
+          if (size < min_stack) size = min_stack;
+#elif PTHREAD_STACK_MIN > 0
+          if (size < PTHREAD_STACK_MIN) size = PTHREAD_STACK_MIN;
+#endif
 #endif
           size = ((size+page_size-1)/page_size)*page_size;